File: backend/app/schemas/transaction.py
import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import Literal, Optional

from pydantic import BaseModel, field_validator


class TransactionCreate(BaseModel):
    category_id: uuid.UUID
    amount: Decimal
    type: Literal["income", "expense"]
    transaction_date: date
    counterparty_name: Optional[str] = None
    memo: Optional[str] = None
    event_type: Optional[str] = None

    @field_validator("amount")
    @classmethod
    def amount_must_be_positive(cls, v: Decimal) -> Decimal:
        if v <= 0:
            raise ValueError("금액은 0보다 커야 합니다")
        return v


class TransactionUpdate(BaseModel):
    category_id: Optional[uuid.UUID] = None
    amount: Optional[Decimal] = None
    type: Optional[Literal["income", "expense"]] = None
    transaction_date: Optional[date] = None
    counterparty_name: Optional[str] = None
    memo: Optional[str] = None
    event_type: Optional[str] = None


class TransactionResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: uuid.UUID
    user_id: uuid.UUID
    category_id: uuid.UUID
    amount: Decimal
    type: str
    transaction_date: date
    counterparty_name: Optional[str]
    memo: Optional[str]
    event_type: Optional[str]
    created_at: datetime


class TransactionListResponse(BaseModel):
    items: list[TransactionResponse]
    total: int
    skip: int
    limit: int

File: frontend/mobile/lib/models/transaction.dart
class Transaction {
  final String id;
  final String userId;
  final String categoryId;
  final double amount;
  final String type; // income | expense
  final DateTime transactionDate;
  final String? counterpartyName;
  final String? memo;
  final String? eventType;
  final DateTime createdAt;

  Transaction({
    required this.id,
    required this.userId,
    required this.categoryId,
    required this.amount,
    required this.type,
    required this.transactionDate,
    this.counterpartyName,
    this.memo,
    this.eventType,
    required this.createdAt,
  });

  factory Transaction.fromJson(Map<String, dynamic> json) => Transaction(
        id: json['id'],
        userId: json['user_id'],
        categoryId: json['category_id'],
        amount: double.parse(json['amount'].toString()),
        type: json['type'],
        transactionDate: DateTime.parse(json['transaction_date']),
        counterpartyName: json['counterparty_name'],
        memo: json['memo'],
        eventType: json['event_type'],
        createdAt: DateTime.parse(json['created_at']),
      );

  bool get isExpense => type == 'expense';
}

class TransactionListResponse {
  final List<Transaction> items;
  final int total;

  TransactionListResponse({required this.items, required this.total});

  factory TransactionListResponse.fromJson(Map<String, dynamic> json) =>
      TransactionListResponse(
        items: (json['items'] as List).map((e) => Transaction.fromJson(e)).toList(),
        total: json['total'],
      );
}
